### PR TITLE
Improve typehints and docs in `diffusers/models`

### LINKED ISSUE #5249: Missing docs in multiple diffusers/models files
Multiple classes such as [this](https://github.com/huggingface/diffusers/blob/78a78515d64736469742e5081337dbcf60482750/src/diffusers/models/unet_2d_blocks.py#L1114) in the models directory are missing documentation or an explanation for what the different parameters mean, and sometimes type hints. Has it been done this way by choice to allow for a clean and direct code view? if not, I would be happy to take this up as an issue and write the docs.

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Fixes #5249.

Improves type hints and docs in the following files:
- [diffusers/models/adapter.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/adapter.py)
- [diffusers/models/attention.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention.py)
- [diffusers/models/lora.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/lora.py)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

@DN6 @yiyixuxu

### CODE REVIEW

**Comment by DN6** [comment]
cc @williamberman @sayakpaul could you take a look here please.

File: src/diffusers/models/adapter.py [hunk 1]
```diff
@@ -367,7 +379,22 @@ def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
 
 
 class AdapterBlock(nn.Module):
-    def __init__(self, in_channels, out_channels, num_res_blocks, down=False):
+    r"""
+    An AdapterBlock is a helper model that contains multiple ResNet-like blocks. It is used in the FullAdapter and
+    FullAdapterXL models.
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:384]:
```suggestion
    An AdapterBlock is a helper model that contains multiple ResNet-like blocks. It is used in the `FullAdapter` and
    `FullAdapterXL` models.
```

File: src/diffusers/models/adapter.py [hunk 2]
```diff
@@ -400,13 +427,21 @@ def forward(self, x):
 
 
 class AdapterResnetBlock(nn.Module):
-    def __init__(self, channels):
+    r"""
+    An AdapterResnetBlock is a helper model that implements a ResNet-like block.
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:431]:
```suggestion
    An `AdapterResnetBlock` is a helper model that implements a ResNet-like block.
```

File: src/diffusers/models/adapter.py [hunk 3]
```diff
@@ -494,13 +548,22 @@ def forward(self, x):
 
 
 class LightAdapterResnetBlock(nn.Module):
-    def __init__(self, channels):
+    """
+    A LightAdapterResnetBlock is a helper model that implements a ResNet-like block with a slightly different
+    architecture than AdapterResnetBlock.
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:553]:
```suggestion
    A `LightAdapterResnetBlock` is a helper model that implements a ResNet-like block with a slightly different
    architecture than `AdapterResnetBlock`.
```

File: src/diffusers/models/lora.py [hunk 4]
```diff
@@ -112,7 +169,7 @@ def __init__(self, *args, lora_layer: Optional[LoRAConv2dLayer] = None, **kwargs
     def set_lora_layer(self, lora_layer: Optional[LoRAConv2dLayer]):
         self.lora_layer = lora_layer
 
-    def _fuse_lora(self, lora_scale=1.0, safe_fusing=False):
+    def _fuse_lora(self, lora_scale: float = 1.0, safe_fusing: bool False):
```
Comment (patrickvonplaten) [inline src/diffusers/models/lora.py:172]:
```suggestion
    def _fuse_lora(self, lora_scale: float = 1.0, safe_fusing: bool = False):
```

**Comment by patrickvonplaten** [comment]
Great work @a-r-r-o-w !

**Pull request merged**